consensus/src/runway/collection.rs

```diff
@@ -0,0 +1,536 @@
+#[derive(Debug, Encode, Decode, Clone)]
+pub struct NewestUnitResponse<H: Hasher, D: Data, S: Signature> {
+    requester: NodeIndex,
+    responder: NodeIndex,
+    unit: Option<UncheckedSignedUnit<H, D, S>>,
+    salt: Salt,
+}
+
+impl<H: Hasher, D: Data, S: Signature> Signable for NewestUnitResponse<H, D, S> {
+    type Hash = Vec<u8>;
+
+    fn hash(&self) -> Self::Hash {
+        self.encode()
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature> crate::Index for NewestUnitResponse<H, D, S> {
+    fn index(&self) -> NodeIndex {
+        self.responder
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature> NewestUnitResponse<H, D, S> {
+    /// Create a newest unit response.
+    pub fn new(
+        requester: NodeIndex,
+        responder: NodeIndex,
+        unit: Option<UncheckedSignedUnit<H, D, S>>,
+        salt: Salt,
+    ) -> Self {
+        NewestUnitResponse {
+            requester,
+            responder,
+            unit,
+            salt,
+        }
+    }
+
+    /// The unit included in this response, if any.
+    pub fn included_unit(&self) -> &Option<UncheckedSignedUnit<H, D, S>> {
+        &self.unit
+    }
+
+    /// The data included in this message, i.e. contents of the unit if any.
+    pub fn included_data(&self) -> Vec<D> {
+        match &self.unit {
+            Some(u) => vec![u.as_signable().data().clone()],
+            None => Vec::new(),
+        }
+    }
+
+    /// Who requested this response.
+    pub fn requester(&self) -> NodeIndex {
+        self.requester
+    }
+}
+
+/// Ways in which a newest unit response might be wrong.
+#[derive(PartialEq, Debug)]
+pub enum Error<H: Hasher, D: Data, S: Signature> {
+    WrongSignature,
+    SaltMismatch(Salt, Salt),
+    InvalidUnit(ValidationError<H, D, S>),
+    ForeignUnit(NodeIndex),
+}
+
+impl<H: Hasher, D: Data, S: Signature> Display for Error<H, D, S> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        use Error::*;
+        match self {
+            WrongSignature => write!(f, "wrong signature"),
+            SaltMismatch(expected, got) => {
+                write!(f, "mismatched salt, expected {}, got {}", expected, got)
+            }
+            InvalidUnit(e) => write!(f, "invalid unit: {}", e),
+            ForeignUnit(id) => write!(f, "unit from node {:?}", id),
+        }
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature> From<ValidationError<H, D, S>> for Error<H, D, S> {
+    fn from(ve: ValidationError<H, D, S>) -> Self {
+        Error::InvalidUnit(ve)
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature> From<SignatureError<NewestUnitResponse<H, D, S>, S>>
+    for Error<H, D, S>
+{
+    fn from(_: SignatureError<NewestUnitResponse<H, D, S>, S>) -> Self {
+        Error::WrongSignature
+    }
+}
+
+/// The status of an ongoing collection.
+#[derive(PartialEq, Debug)]
+pub enum Status {
+    /// Received less than threshold responses.
+    Pending,
+    /// Received at least threshold responses.
```

Comment:
Nit: This comment is not quite correct. It should be "Received at least `threshold - 1` responses" (unless we count ourself as a response) similar is with `Status::Pending`. So maybe we should change those comments or remove adding `keychain.index()` to responders?

Reply:
Mentioned counting the trivial self-response.